src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -116,66 +127,120 @@ ShardManager::AsyncResult< ShardInfo > HSHomeObject::_seal_shard(ShardInfo const
+    req->header()->pg_id = pg_id;
+    req->header()->shard_id = shard_id;
+    req->header()->payload_size = sizeof(shard_info_superblk);
+    req->header()->payload_crc = crc32_ieee(init_crc32, sb_blob.cbytes(), sizeof(shard_info_superblk));
+    req->header()->seal();
+
+    // Similar to create shard - ShardInfo block is persisted on both on header and in data portion.
+    std::memcpy(req->header_extn(), sb_blob.cbytes(), sizeof(shard_info_superblk));
+    req->add_data_sg(std::move(sb_blob));
 
     // replicate this seal shard message to PG members;
-    repl_dev->async_alloc_write(header, sisl::blob{buf_ptr, (uint32_t)msg_size}, value, req);
+    repl_dev->async_alloc_write(req->cheader_buf(), sisl::blob{}, req->data_sgs(), req);
     return req->result();
 }
 
-void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header, homestore::MultiBlkId const& blkids,
+void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& h, homestore::MultiBlkId const& blkids,
                                            shared< homestore::ReplDev > repl_dev,
                                            cintrusive< homestore::repl_req_ctx >& hs_ctx) {
+    repl_result_ctx< ShardManager::Result< ShardInfo > >* ctx{nullptr};
+    if (hs_ctx && hs_ctx->is_proposer) {
+        ctx = boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(hs_ctx).get();
+    }
+
+    auto header = r_cast< const ReplicationMessageHeader* >(h.cbytes());
+    if (header->corrupted()) {
+        LOGW("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::CRC_MISMATCH)); }
+        return;
+    }
+
+#ifdef VADLIDATE_ON_REPLAY
+    sisl::io_blob_safe value_blob(blkids.blk_count() * repl_dev->get_blk_size(), io_align);
+    sisl::sg_list value_sgs;
+    value_sgs.iovs.emplace_back(iovec{.iov_base = value_blob.bytes(), .iov_len = value_blob.size()});
+    value_sgs.size += value_blob.size();
+
+    // Do a read sync read and validate the crc
+    std::error_code err = repl_dev->async_read(blkids, value_sgs, value_blob.size()).get();
+    if (err) {
+        LOGW("failed to read data from homestore blks, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::UNKNOWN)); }
+        return;
+    }
 
-    if (hs_ctx != nullptr) {
-        do_shard_message_commit(lsn, *r_cast< ReplicationMessageHeader* >(const_cast< uint8_t* >(header.cbytes())),
-                                blkids, hs_ctx->key, hs_ctx);
+    if (crc32_ieee(init_crc32, value.cbytes(), value.size()) != header->payload_crc) {
+        // header & value is inconsistent;
+        LOGW("replication message header is inconsistent with value, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::CRC_MISMATCH)); }
         return;
     }
+#endif
```

Comment:
Have this option to do an optional validation of every commit to see if actual data written is same as whats there in journal. We should perhaps make this as run-time option enabled in may be debug build?

Reply:
The normal use case would be to only enable it _after_ some inconsistency was detected right? Do we have an analogous "safe mode" yet?

Reply:
Yes we will enable this dynamic config after some inconsistency. Its also ok to do the same for all debug builds as well. Its good question to ask '_should we tie this config to check as part of "safe mode"_'. Perhaps I can create an issue to track that.